minerl/Malmo/Minecraft/src/main/java/com/microsoft/Malmo/MissionHandlers/ObservationFromDamageImplementation.java

```diff
@@ -0,0 +1,131 @@
+package com.microsoft.Malmo.MissionHandlers;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonPrimitive;
+import com.microsoft.Malmo.MissionHandlerInterfaces.IObservationProducer;
+import com.microsoft.Malmo.Schemas.MissionInit;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.entity.EntityPlayerSP;
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.DamageSource;
+import net.minecraft.util.math.Vec3d;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.entity.living.LivingDeathEvent;
+import net.minecraftforge.event.entity.living.LivingHurtEvent;
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+
+
+import static java.lang.Math.*;
+
+/**
+ * Simple IObservationProducer object that pings out a whole bunch of data.<br>
+ */
+public class ObservationFromDamageImplementation extends HandlerBase implements IObservationProducer
+{
+    private DamageSource damageSource;
+    private float damageAmount;
+    private EntityLivingBase entity;
+    private boolean hasDied = false;
+
+
+	@Override
+	public void prepare(MissionInit missionInit) {
+        MinecraftForge.EVENT_BUS.register(this);
+    }
+
+	@Override
+	public void cleanup() {}
+
+    @SubscribeEvent
+    public void onClientTick(LivingDeathEvent event)
+    {
+        if (event.getEntityLiving().equals(Minecraft.getMinecraft().player)) {
+            if (this.damageSource != null) {
+                System.out.println("Warning overwriting damage source - entity has died!");
+            }
+            this.damageSource = event.getSource();
+            this.entity = event.getEntityLiving();
+            this.hasDied = true;
+        }
+    }
+    
+    @SubscribeEvent
+    public void onDamage(LivingHurtEvent event)
+    {
+        if (event.getEntityLiving().equals(Minecraft.getMinecraft().player)) {
+            if (this.damageSource != null) {
+                System.out.println("Warning skipped damage event - multiple damage events in one tick!");
```

Comment:
I see -- so the reason that landing in this if branch means that there are multiple damage events in one tick is because we expect  `writeObservationsToJSON` to be called every tick, which resets `this.damageSource` to `null`?

If that's true, then could you add that as a comment above the warning log?

Reply:
Mainly to make the implication immediately obvious here

Reply:
Right - this was left as a debugging artifact - I'll clarify in the comments. The issue is I expect that in rare cases you could receive multiple damage events per tick - perhaps I should see if this is possible first

Reply:
```suggestion
                System.out.println("Warning skipped damage event - multiple damage events in one tick!");
                System.out.println("This damage event will be marked as stale=True.")
```